src/relations/tls.py

```diff
@@ -1,292 +1,259 @@
-        self.container.replan()
-        logger.info("TLS enabled.")
-
-    def _unset_tls(self) -> None:
-        """Disable TLS."""
-        for file in [TLS_SSL_KEY_FILE, TLS_SSL_CERT_FILE, TLS_SSL_CONFIG_FILE]:
-            self._remove_file(f"{ROUTER_CONFIG_DIRECTORY}/{file}")
-        # remove tls layer overriding with original layer
-        self.container.add_layer(
-            MYSQL_ROUTER_SERVICE_NAME, self.charm.mysql_router_layer, combine=True
+        self.framework.observe(
+            self._interface.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self._interface.on.certificate_expiring, self._on_certificate_expiring
         )
-        self.container.replan()
-        logger.info("TLS disabled.")
-
-    def _write_content_to_file(
-        self,
-        path: str,
-        content: str,
-        owner: str,
-        group: str,
-        permission: int = 0o640,
-    ) -> None:
-        """Write content to file.
-
-        Args:
-            path: filesystem full path (with filename)
-            content: string content to write
-            owner: file owner
-            group: file group
-            permission: file permission
-        """
-        self.container.push(path, content, permissions=permission, user=owner, group=group)
 
-    def _create_tls_config_file(self) -> None:
-        """Render TLS template directly to file.
+    @property
+    def _relation(self) -> typing.Optional[_Relation]:
+        if not self._charm.model.get_relation(self.NAME):
+            return
+        return _Relation(_charm=self._charm, _interface=self._interface)
 
-        Render and write TLS enabling config file from template.
-        """
-        with open("templates/tls.cnf", "r") as template_file:
-            template = Template(template_file.read())
-            config_string = template.substitute(
-                tls_ssl_key_file=f"{ROUTER_CONFIG_DIRECTORY}/{TLS_SSL_KEY_FILE}",
-                tls_ssl_cert_file=f"{ROUTER_CONFIG_DIRECTORY}/{TLS_SSL_CERT_FILE}",
-            )
-
-        self._write_content_to_file(
-            f"{ROUTER_CONFIG_DIRECTORY}/{TLS_SSL_CONFIG_FILE}",
-            config_string,
-            owner=MYSQL_ROUTER_USER_NAME,
-            group=MYSQL_ROUTER_USER_NAME,
-            permission=0o600,
-        )
+    @property
+    def certificate_saved(self) -> bool:
+        """Whether a TLS certificate is available to use"""
+        if self._relation is None:
+            return False
+        return self._relation.certificate_saved
+
+    def _on_set_tls_private_key(self, event: ops.ActionEvent) -> None:
+        """Handle action to set unit TLS private key."""
+        logger.debug("Handling set TLS private key action")
+        if self._relation is None:
+            event.fail("No TLS relation available.")
+            logger.debug("Unable to set TLS private key: no TLS relation available")
+            return
```

Comment:
If we establish a relationship first
1) certificate will be requested, received, router restarted.
2) we changed the private key to desired one. A certificate will be requested, received, and the router will be restarted.
So we have two restarts in total.

If we allow setting the private key before the relation - we get only one restart.

Reply:
https://github.com/canonical/mysql-router-k8s-operator/pull/51#discussion_r1191122462